[Bug]: latest LiteLLM introduced dependency on "madoka" - which lacks wheels for most environments

### What happened?

Latest LiteLLM versions added dependency on `pondpond` - which in turn depends on `madoka`.
The later lacks wheels for most environments and requires explicit compilation.
Which, for example, fails on "clean" python docker image.

The dependency did not exist in version 1.77.1 - which I've pinned for now.

Here are sample error logs from my docker build:
```
10:18:38 #21 14.47   × Failed to build `madoka==0.7.1`
10:18:38 #21 14.47   ├─▶ The build backend returned an error
10:18:38 #21 14.47   ╰─▶ Call to `setuptools.build_meta:__legacy__.build_wheel` failed (exit
10:18:38 #21 14.47       status: 1)
10:18:38 #21 14.47 
10:18:38 #21 14.47       [stderr]
10:18:38 #21 14.47       error: command 'g++' failed: No such file or directory
10:18:38 #21 14.47 
10:18:38 #21 14.47       hint: This usually indicates a problem with the package or the build
10:18:38 #21 14.47       environment.
10:18:38 #21 14.47   help: `madoka` (v0.7.1) was included because `langroid[litellm]` (v0.59.6)
10:18:38 #21 14.47         depends on `litellm` (v1.77.3) which depends on `pondpond` (v1.4.1)
10:18:39 #21 14.47         which depends on `madoka`
10:18:39 #21 ERROR: process "/bin/sh -c uv pip install --system --no-cache-dir -r requirements.txt" did not complete successfully: exit code: 1
10:18:39 ------
10:18:39  > [management 5/5] RUN uv pip install --system --no-cache-dir -r requirements.txt:
10:18:39 14.47 
10:18:39 14.47       !!
10:18:39 14.47         self._finalize_license_expression()
10:18:39 14.47       error: command 'g++' failed: No such file or directory
10:18:39 14.47 
10:18:39 14.47       hint: This usually indicates a problem with the package or the build
10:18:39 14.47       environment.
10:18:39 ------
```

### Relevant log output

```shell

```

### Are you a ML Ops Team?

No

### What LiteLLM version are you on ?

1.77.3

### Twitter / LinkedIn details

_No response_

Sharing MacOS error for completeness as it is slightly different 

ProductName:		macOS
ProductVersion:		15.6.1
BuildVersion:		24G90


    Hardware Overview:

      Model Name: MacBook Air
      Model Identifier: MacBookAir10,1
      Model Number: MGQN3LL/A
      Chip: Apple M1
      Total Number of Cores: 8 (4 performance and 4 efficiency)
      Memory: 16 GB
      System Firmware Version: 11881.140.96
      OS Loader Version: 11881.140.96

xcode-select version 2410.

```
❯ uv add madoka
Resolved 63 packages in 36ms
  × Failed to build `madoka==0.7.1`
  ├─▶ The build backend returned an error
  ╰─▶ Call to `setuptools.build_meta:__legacy__.build_wheel` failed (exit status: 1)

      [stderr]
      /Users/mjh-ao/Library/Caches/uv/builds-v0/.tmpWRt7Al/lib/python3.12/site-packages/setuptools/dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!

              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:

              License :: OSI Approved :: BSD License

              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************

      !!
        self._finalize_license_expression()
      madoka_wrap.cxx:3059:10: fatal error: 'stdexcept' file not found
       3059 | #include <stdexcept>
            |          ^~~~~~~~~~~
      1 error generated.
      error: command '/usr/bin/c++' failed with exit code 1
```

--- a/litellm/litellm_core_utils/object_pooling.py
+++ b/litellm/litellm_core_utils/object_pooling.py
@@ -16,11 +16,35 @@ Memory Management Strategy:
 
 from typing import Any, Callable, Optional, Type, TypeVar
 
-from pond import Pond, PooledObject, PooledObjectFactory
+try:
+    from pond import Pond, PooledObject, PooledObjectFactory  # type: ignore
+    POND_AVAILABLE = True
+except ImportError:  # pragma: no cover
+    POND_AVAILABLE = False
+    class Pond:  # type: ignore
+        def __init__(self, *args: Any, **kwargs: Any) -> None:
+            pass
+
+        def register(self, *args: Any, **kwargs: Any) -> None:
+            pass
+
+        def borrow(self, *args: Any, **kwargs: Any) -> Any:
+            pass
+
+        def recycle(self, *args: Any, **kwargs: Any) -> None:
+            pass
+
+    class PooledObject:  # type: ignore
+        def __init__(self, keeped_object: Any = None) -> None:
+            self.keeped_object = keeped_object
+
+    class PooledObjectFactory:  # type: ignore
+        def __init__(self, *args: Any, **kwargs: Any) -> None:
+            pass
 
 T = TypeVar('T')
 
-class GenericPooledObjectFactory(PooledObjectFactory):
+class GenericPooledObjectFactory(PooledObjectFactory):  # type: ignore[misc]
     """Generic factory class for creating pooled objects of any type."""
     
     def __init__(
@@ -79,7 +103,7 @@ def get_object_pool(
     time_between_eviction_runs: int = 300,  # Less frequent eviction to maintain high reuse ratio
     eviction_weight: float = 0.3,  # Less aggressive eviction for better reuse
     prewarm_count: int = 5  # Lower pre-warm count to reduce initial memory usage
-) -> Pond:
+) -> Pond | None:
     """Get or create a global object pool instance with balanced eviction-based memory control.
     
     Memory is controlled through moderate eviction to balance reuse ratio and memory usage:
@@ -98,9 +122,13 @@ def get_object_pool(
         prewarm_count: Number of objects to pre-warm the pool with (default: 5)
     
     Returns:
-        Pond instance for the specified object type
+        Pond instance for the specified object type or None if pond is not available
     """
     
+    # If pond is not available, disable pooling gracefully
+    if not POND_AVAILABLE:
+        return None
+
     if pool_name in _pools:
         return _pools[pool_name]
     

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -31,7 +31,7 @@ jinja2 = "^3.1.2"
 aiohttp = ">=3.10"
 pydantic = "^2.5.0"
 jsonschema = "^4.22.0"
-pondpond = "^1.4.1"
+pondpond = {version = "^1.4.1", optional = true}
 numpydoc = {version = "*", optional = true} # used in utils.py
 fastuuid = {version = ">=0.12.0", optional = true}
 
@@ -94,6 +94,7 @@ proxy = [
     "rich",
     "polars",
     "fastuuid",
+    "pondpond",
 ]
 
 extra_proxy = [

--- a/tests/litellm_utils_tests/test_object_pooling.py
+++ b/tests/litellm_utils_tests/test_object_pooling.py
@@ -3,6 +3,7 @@ Simplified tests for object pooling utilities in litellm.
 """
 
 import pytest
+pytest.importorskip("pond")
 
 from litellm.litellm_core_utils.object_pooling import (
     get_object_pool,

